fix(ecs): new arn format not supported (under feature flag)

jumic opened the pull request:
AWS has changed the [ARN format for ECS](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-account-settings.html#ecs-resource-ids). Currently, CDK doesn't return the correct values/ARNs if the new ARN format is used in ECS.

Changed methods:
- `Ec2Service.fromEc2ServiceAttributes()`
- `Ec2Service.fromEc2ServiceArn()`
- `FargateService.fromFargateServiceAttributes()`
- `FargateService.fromFargateServiceArn()`

The logic automatically detects whether the old or new format is used. The format cannot be recognized automatically for tokenized values. Therefore the feature flag `ECS_ARN_FORMAT_INCLUDES_CLUSTER_NAME` is introduced, which controls whether the old or the new format should be used.

In `Ec2Service.fromEc2ServiceAttributes()` and `FargateService.fromFargateServiceAttributes()` an ARN is created. In these methods the feature flag be considered to construct the ARN in the correct format.

Closes #16634.
Closes #18137.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 3223be3723bf6fa9d76312e0e24a5b9907439084
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2a698b20-c05e-45d9-aa41-024e768fb23f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

madeline-k commented on packages/@aws-cdk/aws-ecs/test/ec2/ec2-service.test.ts:
```diff
@@ -3280,7 +3282,106 @@ describe('ec2 service', () => {
+          // THEN
+          expect(stack.resolve(service.serviceArn)).toEqual({ Ref: 'ARN' });
+          expect(stack.resolve(service.serviceName)).toEqual({
+            'Fn::Select': [
+              1,
+              {
+                'Fn::Split': [
+                  '/',
+                  {
+                    'Fn::Select': [
+                      5,
+                      {
+                        'Fn::Split': [
+                          ':',
+                          { Ref: 'ARN' },
+                        ],
+                      },
+                    ],
```
Oof, this is very hard to determine if this is correct (although I am pretty sure it is). And unfortunately, I don't think there is another way to implement this to make the resulting template more readable.

jumic replied:
Should I add integration tests that show the result using CloudFormation Outputs? I tested the changes using this code:

```ts
const cluster = new ecs.Cluster(this, 'Cluster', {
  vpc,
});

const param = new CfnParameter(this, 'Param', {
  default: 'arn:aws:ecs:us-west-2:123456789012:service/my-http-service',
})

const fromFargateServiceArn = ecs.FargateService.fromFargateServiceArn(this, 'fromFargateServiceArn', param.valueAsString);
new CfnOutput(this, 'fromFargateServiceArnOutput', {
  value: fromFargateServiceArn.serviceName,
});

const fromEc2ServiceArn = ecs.Ec2Service.fromEc2ServiceArn(this, 'fromEc2ServiceArn', param.valueAsString);
new CfnOutput(this, 'fromEc2ServiceArnOutPut', {
  value: fromEc2ServiceArn.serviceName,
});

const fromEc2ServiceAttributes = ecs.Ec2Service.fromEc2ServiceAttributes(this, 'fromEc2ServiceAttributes', {
  cluster,
  serviceArn: param.valueAsString,
});
new CfnOutput(this, 'fromEc2ServiceAttributesOutput', {
  value: fromEc2ServiceAttributes.serviceName,
});

const fromFargateServiceAttributes = ecs.FargateService.fromFargateServiceAttributes(this, 'fromFargateServiceAttributes', {
  cluster,
  serviceArn: param.valueAsString,
});
new CfnOutput(this, 'fromFargateServiceAttributesOutput', {
  value: fromFargateServiceAttributes.serviceName,
});
```

New ARN Format and feature flag enabled (`CfnParameter`: `arn:aws:ecs:us-west-2:123456789012:service/cluster-name/my-http-service`) - Output:

```
MyStack.fromEc2ServiceArnOutPut = my-http-service
MyStack.fromEc2ServiceAttributesOutput = my-http-service
MyStack.fromFargateServiceArnOutput = my-http-service
MyStack.fromFargateServiceAttributesOutput = my-http-service
```

Old ARN Format and feature flag disabled (`CfnParameter`: `arn:aws:ecs:us-west-2:123456789012:service/my-http-service`) - Output:
```
MyStack.fromEc2ServiceArnOutPut = my-http-service
MyStack.fromEc2ServiceAttributesOutput = my-http-service
MyStack.fromFargateServiceArnOutput = my-http-service
MyStack.fromFargateServiceAttributesOutput = my-http-service
```

madeline-k reviewed (changes requested):
Thanks for contributing this fix, @jumic! This looks great, I just have a couple questions/comments.

jumic commented on packages/@aws-cdk/aws-ecs/lib/base/from-service-attributes.ts:
```diff
@@ -32,22 +37,38 @@ export function fromServiceAtrributes(scope: Construct, id: string, attrs: Servi
   }
 
+  const featureConstruct = new CoreConstruct(scope, randomString());
+  const newArnFormat = FeatureFlags.of(featureConstruct).isEnabled(ECS_ARN_FORMAT_INCLUDES_CLUSTER_NAME);
```
I tried `FeatureFlags.of(scope)` first because it should be the preferred solution. Unfortunately, it causes this error:

```
Argument of type 'import("/Users/julian/dev/aws-cdk/node_modules/constructs/lib/construct").Construct' is not assignable to parameter of type 'import("/Users/julian/dev/aws-cdk/packages/@aws-cdk/core/lib/construct-compat").Construct'.
Type 'Construct' is missing the following properties from type 'Construct': node, validate, prepare, synthesizets(2345)
```

madeline-k replied:
Since FeatureFlags expects a `Construct` from the core library, I think we can just switch to using that one in the `fromServiceAttributes()` function. I pushed this change to your branch, waiting for the build to see if this has other impacts.

jumic commented on packages/@aws-cdk/aws-ecs/lib/ec2/ec2-service.ts:
```diff
@@ -128,9 +129,25 @@ export class Ec2Service extends BaseService implements IEc2Service {
+      }
     }
-    return new Import(scope, id);
+    return new Import();
```
Previously, passing the parameters was necessary because they were passed to the super class. The new implementation contains a constructor. In this case, the `super()` call has to be implemented in the constructor. Therefore, the parameters are not necessary anymore.

If you prefer the existing code (`return new Import(scope, id);`), I could implement it too. For example:

```ts
  constructor(importScope: Construct, importId: string) {
    super(importScope, importId);
    // ...
  }
}
return new Import(scope, id);
```

However, we have to use new attribute names. If I use `scope` instead of `importScope`, the following error is displayed:`'scope' is already declared in the upper scope.eslint@typescript-eslint/no-shadow`.

Which option do you prefer?

jumic replied:
No constructor is used anymore --> done.

tobytipton commented:
Implementing this as a feature flag impacts both existing and new services.  The use case would be if you have an service which was created with the old Arn format and are extending your CDK to create a addition new ECS service which would use the new Arn format.  The feature flag would need to be set for the new ECS service with the new Arn Format.  But it would impact the existing service as well.

Maybe a property would be more desirable?

jumic commented on packages/@aws-cdk/aws-ecs/lib/ec2/ec2-service.ts:
```diff
@@ -128,9 +129,25 @@ export class Ec2Service extends BaseService implements IEc2Service {
+            this.serviceName = Fn.select(2, lastComponents);
+          } else {
+            this.serviceName = resourceName;
+          }
```
I moved the code to a new function `extractServiceNameFromArn` in `from-service-attributes.ts`.

jumic commented on packages/@aws-cdk/aws-ecs/lib/ec2/ec2-service.ts:
```diff
@@ -128,9 +129,25 @@ export class Ec2Service extends BaseService implements IEc2Service {
+          }
+        } else {
+          const resourceNameSplit = resourceName.split('/');
+          this.serviceName = resourceNameSplit.length === 1 ? resourceName : resourceNameSplit[1];
```
I moved the code to a new function `extractServiceNameFromArn` in `from-service-attributes.ts`.

jumic commented:
> Implementing this as a feature flag impacts both existing and new services. The use case would be if you have an service which was created with the old Arn format and are extending your CDK to create a addition new ECS service which would use the new Arn format. The feature flag would need to be set for the new ECS service with the new Arn Format. But it would impact the existing service as well.
> 
> Maybe a property would be more desirable?

You're right. The solution in this PR doesn't provide an option to use the new and the old format in a single stack at the same time. I also proposed a new property in my [original comment](https://github.com/aws/aws-cdk/issues/16634#issuecomment-943719465).

@tobytipton Do you have the requirement to mix the old and the new format in a single stack for tokenized ARNs? In this case, we could think about implementing a property. If there is no requirement, I wouldn't support this because we don't know if somebody will ever use it.

In any case, the feature flag should be implemented. This simplifies the use of the method, since the `newFormat` property does not have to be defined every time in new stacks.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f5a35603a34e155e83777b93266f7773eed53f17
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1b390a16-2c8f-49eb-bec9-79925e3441e6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: b5c9bf52e8c3d7f74a4faf42955dd9bdb701e471
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bfa23f28-9d31-4d8a-b812-99780a1e0a0f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jumic commented:
> Here is the scenario I can think of for having new and old formats in the token. Prior to the change of the ARN format Sept 2020 a ECS/Fargate service is generated using CDK, that would have the old ARN format. Today (after the new ARN format is the default) I want to add an additional ECS/Fargate service to the same stack or application which adds on to the original CDK, that new service would have the new ARN format.
> 
> I see in your tests you are having the token coming from a CFN parameter, but still you could have a new and old format at the same time.
> 
> So possibly there does need to be both a feature flag and a property?

I totally agree. In this case, we need both. A property and a feature flag.

@madeline-k Should I add a property to support parsing the old and the new format from a tokenized ARN in a single stack?
So far I haven't seen a feature request / bug report that requests this functionality. So I'm not sure we should really add this feature. Of course we can add it later if requested.

madeline-k commented:
Hey @tobytipton  and @jumic, thanks for working on this issue! There is currently some overlap between this PR and https://github.com/aws/aws-cdk/pull/18382. 

What I'm thinking is this PR should fix the issues with `ecs.Fargate/Ec2Service.fromXXX()` producing an `IService` with the wrong values for `serviceArn` and `serviceName`. And #18382 should address only that the tokenized ARN for a newly created service does not have region/account info. This means that this PR will need to cover some of the test cases that @tobytipton included in #18382.  Let me know if that makes sense!

madeline-k commented:
@jumic, regarding:
> @madeline-k Should I add a property to support parsing the old and the new format from a tokenized ARN in a single stack? So far I haven't seen a feature request / bug report that requests this functionality. So I'm not sure we should really add this feature. Of course we can add it later if requested.

I'd say if this is something that we can support later, and this implementation is not going to be closing a door to that, then we should no implement the property in this PR, and keep the scope smaller.

jumic commented:
@madeline-k Yes, this makes totally sense. Keeping the pull requests small was also my original idee when I asked to create a separate issue for the region/account topic.

> I'd say if this is something that we can support later, and this implementation is not going to be closing a door to that, then we should no implement the property in this PR, and keep the scope smaller.

We can add this feature later if it will be requested. The current logic must not be changed. Therefore, we are free to add it later if necessary.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a5343800971bc4c7c7757ea16e72548e3811d689
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9756e074-9f8b-4c60-aedc-ad815bb35ac1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

madeline-k reviewed (dismissed):
Thanks for making this contribution @jumic! Sorry for the delay in getting back to reviewing this. To summarize what has already been said on this thread: this change will use the new ARN format when the `ECS_ARN_FORMAT_INCLUDES_CLUSTER_NAME` feature flag is enabled. Since feature flags operate at the stack level, this means the old and new ARN formats cannot both be used in a single stack. If there is a need to be able to use both the old and new ARN format in a single stack, then we can open an issue for that and add a property to control the ARN format at the construct level. 

The last change that is needed here is an update to the README to explain the feature flag, and explain the restriction about not being able to use the old and new formats within one stack. @jumic, if you are able to include that that would be great!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 30546e66fb41316cf307e49d0911248f4e3af567
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6b54e7fe-283f-429c-a594-d67d5f2d76d3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: dc8bcff78c1452f67d7506d8f690780e5d2503c0
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ce77b333-1ab3-46a2-8f2b-2ea869a576f9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 010502ef225ecaa56f90a28ba0ba36afab1e3c56
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7ac397cd-0146-4444-82b0-07a53d4a95bc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 2a2a047964e52e5e8be77d57e38c35198a942280
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6465d0f8-20d4-4970-8ade-e29d72c6ce01%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 8f672955050d4e400783f0f0886d90ba7f826b4f
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=971ee8aa-8a76-4d57-bfd4-4bfcf3291305%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 6a05966c6b0821ba799aed36d0f340e73254333f
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0bc4b947-6c0c-4232-b785-85806ad89833%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jumic commented:
I added the new feature flag to the README file. 
In addition, I changed the `Construct` import to its original value to resolve the compilation failures.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: dd295d8cea7da9027ea09b13fcab37b2ff20b357
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a89c5fe4-e0b0-4d4e-8bb3-559e0bf33753%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 92136f2f06c48b8cc18d1292592372881e4f7263
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a9562742-d1b2-4331-b632-f7af67550de3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4db80e214e3d05b76c2a7fd63edc9fbadc06b0e1
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=83721ef3-6269-4e7b-bac9-a7491c89d549%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented:
@jumic I tried to do some quick fixes because of the branch change but there are still some test failures to look at. Could you please take a look at those?

jumic commented:
@TheRealAmazonKendra I analyzed the test failures. The test are not executed correctly because there is an issue with the feature flag logic in CDK v2. I opened issue #20890. I will check this PR again after the issue is resolved.

TheRealAmazonKendra commented:
Ah, the feature flag issue strikes again. I'm actually working on this issue. In the meantime, the tests can still be done by manually adding the feature flag to the app in the tests instead of using these testing functions. If you feel any urgency about getting this merged, that would be my suggestion as a path forward. Otherwise, I'll update when we have a path forward for the feature flag fix.

TheRealAmazonKendra commented:
Just in case you don't want to wait on the feature flag changes, here is a model for testing with and without the feature flag enabled https://github.com/aws/aws-cdk/pull/21021/files.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9caa1a560e17b449a1078a39f8e92756e599eb18
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=92327bcf-c629-4989-8914-138d70de0361%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 764d034e873fc14d492a1ce75fcd761d1a684ba3
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=de1bd685-8480-46b4-a42a-8b1066d8ce23%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jumic commented:
@TheRealAmazonKendra I added the suggested fix for the current feature flag issue.

TheRealAmazonKendra commented on packages/@aws-cdk/aws-ecs/README.md:
```diff
@@ -703,6 +703,37 @@ There are two higher-level constructs available which include a load balancer fo
+
+Since AWS has changed the [ARN format for ECS](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-account-settings.html#ecs-resource-ids), 
+feature flag `@aws-cdk/aws-ecs:arnFormatIncludesClusterName` must be enabled to use the new ARN format.
+The feature flag changes behavior for the entire CDK project. Therefore it is not possible to mix the old and the new format in one CDK project.
```
This is not strictly true, but we generally recommend to not do this inline and have zero (I think?) instructions on how to do it so I don't think I mind this note here.

TheRealAmazonKendra commented:
Thanks for all your effort on this and for bearing with us on the feature flag issues! Glad we could get this one merged!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: f88e7fb522765f3592bd375a420a8a42999e9448
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4e36b9d0-1dfd-4224-8d70-0653bc141297%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
